src/univariate/continuous/normal.jl

```diff
@@ -85,163 +85,10 @@ end
-        z = zval(Normal(μ, σ), x)
-    end
-    return _normlogpdf(z) - log(σ)
-end
+# Use Julia implementations in StatsFuns
+@_delegate_statsfuns Normal norm μ σ
 
-# pdf
-_normpdf(z::Real) = exp(-abs2(z)/2) * invsqrt2π
-
-function pdf(d::Normal, x::Real)
-    μ, σ = d.μ, d.σ
-    if iszero(σ)
-        if x == μ
-            z = zval(Normal(μ, one(σ)), x)
-        else
-            z = zval(d, x)
-            σ = one(σ)
-        end
-    else
-        z = zval(Normal(μ, σ), x)
-    end
-    return _normpdf(z) / σ
-end
-
-# logcdf
-function _normlogcdf(z::Real)
-    if z < -one(z)
-        return log(erfcx(-z * invsqrt2)/2) - abs2(z)/2
-    else
-        return log1p(-erfc(z * invsqrt2)/2)
-    end
-end
-
-function logcdf(d::Normal, x::Real)
-    if iszero(d.σ) && x == d.μ
-        z = zval(Normal(zero(d.μ), d.σ), one(x))
-    else
-        z = zval(d, x)
-    end
-    return _normlogcdf(z)
-end
-
-# logccdf
-function _normlogccdf(z::Real)
-    if z > one(z)
-        return log(erfcx(z * invsqrt2)/2) - abs2(z)/2
-    else
-        return log1p(-erfc(-z * invsqrt2)/2)
-    end
-end
-
-function logccdf(d::Normal, x::Real)
-    if iszero(d.σ) && x == d.μ
-        z = zval(Normal(zero(d.μ), d.σ), one(x))
-    else
-        z = zval(d, x)
-    end
-    return _normlogccdf(z)
-end
-
-# cdf
-_normcdf(z::Real) = erfc(-z * invsqrt2)/2
-
-function cdf(d::Normal, x::Real)
-    if iszero(d.σ) && x == d.μ
-        z = zval(Normal(zero(d.μ), d.σ), one(x))
-    else
-        z = zval(d, x)
-    end
-    return _normcdf(z)
-end
-
-# ccdf
-_normccdf(z::Real) = erfc(z * invsqrt2)/2
-
-function ccdf(d::Normal, x::Real)
-    if iszero(d.σ) && x == d.μ
-        z = zval(Normal(zero(d.μ), d.σ), one(x))
-    else
-        z = zval(d, x)
-    end
-    return _normccdf(z)
-end
-
-# quantile
-function quantile(d::Normal, p::Real)
-    # Promote to ensure that we don't compute erfcinv in low precision and then promote
-    _p, _μ, _σ = map(float, promote(p, d.μ, d.σ))
-    q = xval(d, -erfcinv(2*_p) * sqrt2)
-    if isnan(_p)
-        return oftype(q, _p)
-    elseif iszero(_σ)
-        # Quantile not uniquely defined at p=0 and p=1 when σ=0
-        if iszero(_p)
-            return oftype(q, -Inf)
-        elseif isone(_p)
-            return oftype(q, Inf)
-        else
-            return oftype(q, _μ)
-        end
-    end
-    return q
-end
-
-# cquantile
-function cquantile(d::Normal, p::Real)
-    # Promote to ensure that we don't compute erfcinv in low precision and then promote
-    _p, _μ, _σ = map(float, promote(p, d.μ, d.σ))
-    q = xval(d, erfcinv(2*_p) * sqrt2)
-    if isnan(_p)
-        return oftype(q, _p)
-    elseif iszero(_σ)
-        # Quantile not uniquely defined at p=0 and p=1 when σ=0
-        if iszero(_p)
-            return oftype(q, Inf)
-        elseif isone(_p)
-            return oftype(q, -Inf)
-        else
-            return oftype(q, _μ)
-        end
-    end
-    return q
-end
+gradlogpdf(d::Normal, x::Real) = -(x - d.μ) / d.σ^2
```

Comment:
Perhaps
```suggestion
gradlogpdf(d::Normal, x::Real) = (d.μ - x) / abs2(d.σ)
```
?

Reply:
I think `abs2` or `^` should not matter, last time I checked they compiled to exactly the same thing. Or do you know any reason why it should be preferred? I'll definitely simplify the calculations though, I just plugged in the definition of `zval` :+1:

Reply:
```suggestion
gradlogpdf(d::Normal, x::Real) = (d.μ - x) / d.σ^2
```

Reply:
> I think `abs2` or `^` should not matter, last time I checked they compiled to exactly the same thing. Or do you know any reason why it should be preferred?

While they do compile to the same thing, `abs2` is more informative for symbolics systems or an AD. e.g. ForwardDiff and ChainRules ADs will end up adding two floats with `abs2` vs multiplying two floats with `^`:

```julia
julia> using ForwardDiff

julia> @code_typed ForwardDiff.derivative(abs2, 3.5)
CodeInfo(
1 ─ %1 = Base.add_float(x, x)::Float64
│   %2 = Base.mul_float(1.0, %1)::Float64
└──      return %2
) => Float64

julia> @code_typed ForwardDiff.derivative(x -> x^2, 3.5)
CodeInfo(
1 ─ %1 = Base.mul_float(2.0, x)::Float64
│   %2 = Base.mul_float(1.0, %1)::Float64
└──      return %2
) => Float64
```

It's not a difference that I've been able to make appear in benchmarks, but this is why I generally use `abs2`.

Reply:
Interesting, I think one would just have to handle `Val{1}` and `Val{2}` cases in https://github.com/JuliaDiff/ForwardDiff.jl/blob/29332b12075e66e056843d3b4207d856f98772b3/src/dual.jl#L506-L516 without the multiplication in https://github.com/JuliaDiff/ForwardDiff.jl/blob/29332b12075e66e056843d3b4207d856f98772b3/src/dual.jl#L513. Due to `literal_pow` `abs2` and `^ 2` should be equally informative.

Reply:
I'll stick with `^2` for now. As mentioned, it is equally informative as `abs2`, and ForwardDiff even contains specialized derivatives for it. The difference here just comes from the fact that the implementation for `literal_pow` uses `2 * x` whereas DiffRules decided to implement the derivative of `abs2` with `x + x` (https://github.com/JuliaDiff/DiffRules.jl/blob/7b1c31ed5b026ea445fd42c9674391a45dc9c536/src/rules.jl#L12).

In fact, ChainRules also uses multiplication for `abs2`, as can seen by the rules (https://github.com/JuliaDiff/ChainRules.jl/blob/13a362c4b5069c6a83f936d9d9d8821e566232bb/src/rulesets/Base/fastmath_able.jl#L84-L94) or running this example:
```julia
julia> using ChainRules

julia> @code_typed last(ChainRules.rrule(abs2, 3.5))(1.0)
CodeInfo(
1 ─ %1 = Base.mul_float(2.0, ΔΩ)::Float64
│   %2 = Core.getfield(#self#, :z)::Float64
│   %3 = Base.mul_float(%1, %2)::Float64
│   %4 = Core.tuple($(QuoteNode(ChainRulesCore.NoTangent())), %3)::Tuple{ChainRulesCore.NoTangent, Float64}
└──      return %4
) => Tuple{ChainRulesCore.NoTangent, Float64}

julia> @code_typed ChainRules.frule((ChainRules.NoTangent(), 0.0), abs2, 3.5)
CodeInfo(
1 ─ %1 = Base.getfield(@_2, 2)::Float64
│   %2 = Base.mul_float(z, z)::Float64
│   %3 = Base.mul_float(z, %1)::Float64
│   %4 = Base.mul_float(2.0, %3)::Float64
│   %5 = Core.tuple(%2, %4)::Tuple{Float64, Float64}
└──      return %5
) => Tuple{Float64, Float64}
```
And CR also implements rules for `literal_pow` (https://github.com/JuliaDiff/ChainRules.jl/blob/13a362c4b5069c6a83f936d9d9d8821e566232bb/src/rulesets/Base/base.jl#L184-L200):
```julia
julia> @code_typed last(ChainRules.rrule(Base.literal_pow, ^, 3.5, Val(2)))(1.0)
CodeInfo(
1 ─ %1 = Core.getfield(#self#, :x)::Float64
│   %2 = Base.mul_float(2.0, %1)::Float64
│   %3 = Base.mul_float(%2, dy)::Float64
│   %4 = Core.tuple($(QuoteNode(ChainRulesCore.NoTangent())), $(QuoteNode(ChainRulesCore.NoTangent())), %3, $(QuoteNode(ChainRulesCore.NoTangent())))::Tuple{ChainRulesCore.NoTangent, ChainRulesCore.NoTangent, Float64, ChainRulesCore.NoTangent}
└──      return %4
) => Tuple{ChainRulesCore.NoTangent, ChainRulesCore.NoTangent, Float64, ChainRulesCore.NoTangent}

julia> @code_typed ChainRules.frule((ChainRules.NoTangent(), ChainRules.NoTangent(), 0.0, ChainRules.NoTangent()), Base.literal_pow, ^, 3.5, Val(2))
CodeInfo(
1 ─ %1 = Base.getfield(@_2, 3)::Float64
│   %2 = Base.mul_float(x, x)::Float64
│   %3 = Base.mul_float(2.0, x)::Float64
│   %4 = Base.mul_float(%3, %1)::Float64
│   %5 = Core.tuple(%2, %4)::Tuple{Float64, Float64}
└──      return %5
) => Tuple{Float64, Float64}
```

So I think it's fine to keep the more natural `^2`. I guess the compiler could decide to generate the same code for `x + x` and `2 * x` but this is something that we shouldn't have to take care in this package, I think.